File: backend/src/main/java/com/example/backend/model/RoomType.java
package com.example.backend.model;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "room_types")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class RoomType {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "room_type_id")
    private UUID id;

    @Column(name = "name", nullable = false, length = 100)
    private String name;

    @Column(name = "capacity", nullable = false)
    private Integer capacity;

    @Column(name = "size_sqm")
    private Integer sizeSqm;

    @Column(name = "total_rooms", nullable = false)
    private Integer totalRooms;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Builder.Default
    @Column(name = "is_available")
    private Boolean isAvailable = true;

    @Builder.Default
    @Column(name = "is_active")
    private Boolean isActive = true;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    // N:1 relationship with Hotel
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "hotel_id", nullable = false)
    private Hotel hotel;

    // 1: N relationship with RoomAmenity
    @OneToMany(mappedBy = "roomType", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<RoomAmenity> amenities;

    // 1: N relationship with Room
    @OneToMany(mappedBy = "roomType", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Room> rooms;

    // 1: N relationship with BaseRate
    @OneToMany(mappedBy = "roomType", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<BaseRate> baseRates;

    // 1: N relationship with DailyOverride
    @OneToMany(mappedBy = "roomType", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<DailyOverride> dailyOverrides;

    // 1: N relationship with BookingRoom
    @OneToMany(mappedBy = "roomType", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<BookingRoom> bookingRooms;
}

File: frontend/src/types/roomType.ts
import type { BaseRateResponse } from "./baseRate";
import type { BaseEntity, BaseFilterRequest } from "./common";
import type { DailyOverrideResponse } from "./dailyOverride";
import type { RoomResponse } from "./room";
import type { RoomAmenityResponse } from "./roomAmenity";

export interface RoomTypeResponse extends BaseEntity {
  name: string;
  capacity: number;
  sizeSqm: number;
  totalRooms: number;
  description?: string;
  isAvailable: boolean;
  rooms: RoomResponse[];
  amenities: RoomAmenityResponse[];
  baseRates: BaseRateResponse[];
  dailyOverrides: DailyOverrideResponse[];
}

export interface RoomTypeListResponse extends BaseEntity {
  name: string;
  capacity: number;
  sizeSqm: number;
  totalRooms: number;
  description?: string;
  isAvailable: boolean;
  baseRates: BaseRateResponse[];
  dailyOverrides: DailyOverrideResponse[];
}

export interface RoomTypeRequest {
  hotelId: string;
  name: string;
  capacity: number;
  sizeSqm: number;
  totalRooms: number;
  description?: string;
}

export interface RoomTypeUpdateRequest {
  name?: string;
  capacity?: number;
  sizeSqm?: number;
  totalRooms?: number;
  description?: string;
  hotelId: string;
}

export interface RoomTypeFilterRequest extends BaseFilterRequest {
  hotelId?: string;
  name?: string;
  capacity?: number;
  minSize?: number;
  maxSize?: number;
  isAvailable?: boolean;
}
